File: impl/sdk/rs/resonate/src/codec.rs
use std::sync::Arc;

use base64::engine::general_purpose::STANDARD as BASE64;
use base64::Engine;
use serde::de::DeserializeOwned;
use serde::{Deserialize, Serialize};

use crate::error::{Error, Result};
use crate::types::{PromiseRecord, Value};

/// Encryption trait for codec (default: no-op).
pub trait Encryptor: Send + Sync {
    fn encrypt(&self, data: &[u8]) -> Result<Vec<u8>>;
    fn decrypt(&self, data: &[u8]) -> Result<Vec<u8>>;
}

/// No-op encryptor (passthrough).
pub struct NoopEncryptor;
impl Encryptor for NoopEncryptor {
    fn encrypt(&self, data: &[u8]) -> Result<Vec<u8>> {
        Ok(data.to_vec())
    }
    fn decrypt(&self, data: &[u8]) -> Result<Vec<u8>> {
        Ok(data.to_vec())
    }
}

/// Handles encoding/decoding of values for the durability boundary.
///
/// Encode: Rust value → JSON → encrypt → base64 → Value { headers, data }
/// Decode: Value { headers, data } → base64 → decrypt → JSON → Rust value
#[derive(Clone)]
pub struct Codec {
    encryptor: Arc<dyn Encryptor>,
}

impl std::fmt::Debug for Codec {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("Codec").finish()
    }
}

impl Codec {
    pub fn new(encryptor: Arc<dyn Encryptor>) -> Self {
        Self { encryptor }
    }

    /// Encode a serializable value into the wire format.
    pub fn encode(&self, value: &impl Serialize) -> Result<Value> {
        let json_val = serde_json::to_value(value)?;
        if json_val.is_null() {
            return Ok(Value {
                headers: None,
                data: Some(serde_json::Value::String(String::new())),
            });
        }
        let json_str = serde_json::to_string(&json_val)?;
        let encrypted = self.encryptor.encrypt(json_str.as_bytes())?;
        let b64 = BASE64.encode(&encrypted);
        Ok(Value {
            headers: None,
            data: Some(serde_json::Value::String(b64)),
        })
    }

    /// Decode a wire-format value back into a Rust type.
    pub fn decode<T: DeserializeOwned>(&self, value: &Value) -> Result<Option<T>> {
        let data = value.data_as_ref();
        let s = match data {
            serde_json::Value::String(s) if s.is_empty() => return Ok(None),
            serde_json::Value::String(s) => s,
            serde_json::Value::Null => return Ok(None),
            _ => return Err(Error::DecodingError("expected string or null data".into())),
        };
        let bytes = BASE64.decode(s)?;
        let decrypted = self.encryptor.decrypt(&bytes)?;
        let json_str = String::from_utf8(decrypted)?;
        let decoded: T = serde_json::from_str(&json_str)?;
        Ok(Some(decoded))
    }

    /// Decode an entire PromiseRecord's param and value fields, consuming the original.
    pub fn decode_promise(&self, promise: PromiseRecord) -> Result<PromiseRecord> {
        let decoded_param_data: serde_json::Value = self
            .decode(&promise.param)?
            .unwrap_or(serde_json::Value::Null);
        let decoded_value_data: serde_json::Value = self
            .decode(&promise.value)?
            .unwrap_or(serde_json::Value::Null);

        Ok(PromiseRecord {
            id: promise.id,
            state: promise.state,
            timeout_at: promise.timeout_at,
            param: Value {
                headers: promise.param.headers,
                data: Some(decoded_param_data),
            },
            value: Value {
                headers: promise.value.headers,
                data: Some(decoded_value_data),
            },
            tags: promise.tags,
            created_at: promise.created_at,
            settled_at: promise.settled_at,
        })
    }

    /// Decode a promise from a raw JSON value (as returned by the network).
    /// Parses the JSON into a PromiseRecord, then decodes param/value fields.
    pub fn decode_promise_from_json(&self, json: &serde_json::Value) -> Result<PromiseRecord> {
        let record: PromiseRecord = PromiseRecord::deserialize(json)
            .map_err(|e| Error::DecodingError(format!("invalid promise JSON: {}", e)))?;
        self.decode_promise(record)
    }
}

/// Encode an error for durable storage.
pub fn encode_error(err: &Error) -> serde_json::Value {
    serde_json::json!({
        "__type": "error",
        "message": err.to_string(),
    })
}

/// Decode an entire PromiseRecord's param and value fields, returning decoded copies.
/// (Public alias for testing convenience.)
impl Codec {
    /// Check if a string is valid base64.
    pub fn is_valid_base64(s: &str) -> bool {
        BASE64.decode(s).is_ok()
    }
}

/// Deserialize an error value from a rejected promise.
pub fn deserialize_error(value: serde_json::Value) -> Error {
    if let Some(obj) = value.as_object() {
        if let Some(msg) = obj.get("message").and_then(|m| m.as_str()) {
            return Error::Application {
                message: msg.to_string(),
            };
        }
    }
    Error::Application {
        message: format!("unknown error: {}", value),
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::types::{PromiseRecord, PromiseState};
    use std::collections::HashMap;

    fn codec() -> Codec {
        Codec::new(Arc::new(NoopEncryptor))
    }

    // ── encode/decode roundtrip for primitives ──────────────────────

    #[test]
    fn roundtrip_integer() {
        let c = codec();
        let encoded = c.encode(&42).unwrap();
        let decoded: Option<i64> = c.decode(&encoded).unwrap();
        assert_eq!(decoded, Some(42));
    }

    #[test]
    fn roundtrip_string() {
        let c = codec();
        let encoded = c.encode(&"hello").unwrap();
        let decoded: Option<String> = c.decode(&encoded).unwrap();
        assert_eq!(decoded, Some("hello".to_string()));
    }

    #[test]
    fn roundtrip_bool() {
        let c = codec();
        let encoded = c.encode(&true).unwrap();
        let decoded: Option<bool> = c.decode(&encoded).unwrap();
        assert_eq!(decoded, Some(true));
    }

    // ── encode/decode roundtrip for objects ──────────────────────────

    #[test]
    fn roundtrip_object() {
        let c = codec();
        let obj = serde_json::json!({"func": "f", "args": [1, "two"]});
        let encoded = c.encode(&obj).unwrap();
        let decoded: Option<serde_json::Value> = c.decode(&encoded).unwrap();
        assert_eq!(decoded, Some(obj));
    }

    // ── encode/decode roundtrip for arrays ──────────────────────────

    #[test]
    fn roundtrip_array() {
        let c = codec();
        let arr = serde_json::json!([1, 2, 3]);
        let encoded = c.encode(&arr).unwrap();
        let decoded: Option<serde_json::Value> = c.decode(&encoded).unwrap();
        assert_eq!(decoded, Some(arr));
    }

    // ── encode None/unit produces empty data ────────────────────────

    #[test]
    fn encode_null_produces_empty_data() {
        let c = codec();
        let encoded = c.encode(&serde_json::Value::Null).unwrap();
        assert_eq!(
            encoded.data_or_null(),
            serde_json::Value::String(String::new())
        );

        let decoded: Option<serde_json::Value> = c.decode(&encoded).unwrap();
        assert!(decoded.is_none());
    }

    // ── encode produces base64 string ───────────────────────────────

    #[test]
    fn encode_produces_valid_base64() {
        let c = codec();
        let encoded = c.encode(&"hello").unwrap();
        let data_str = encoded.data_or_null();
        let data_str = data_str.as_str().unwrap();
        assert!(Codec::is_valid_base64(data_str));

        // Decoding base64 should produce valid JSON
        let bytes = BASE64.decode(data_str).unwrap();
        let json_str = String::from_utf8(bytes).unwrap();
        let _: serde_json::Value = serde_json::from_str(&json_str).unwrap();
    }

    // ── decode_promise decodes both param and value ─────────────────

    #[test]
    fn decode_promise_decodes_param_and_value() {
        let c = codec();
        let param_encoded = c.encode(&serde_json::json!({"func": "f"})).unwrap();
        let value_encoded = c.encode(&serde_json::json!({"result": 42})).unwrap();

        let record = PromiseRecord {
            id: "test".to_string(),
            state: PromiseState::Resolved,
            timeout_at: 0,
            param: param_encoded,
            value: value_encoded,
            tags: HashMap::new(),
            created_at: 0,
            settled_at: Some(1),
        };

        let decoded = c.decode_promise(record).unwrap();
        assert_eq!(
            decoded.param.data_or_null(),
            serde_json::json!({"func": "f"})
        );
        assert_eq!(
            decoded.value.data_or_null(),
            serde_json::json!({"result": 42})
        );
    }

    // ── decode invalid base64 returns error ─────────────────────────

    #[test]
    fn decode_invalid_base64_returns_error() {
        let c = codec();
        let bad_value = Value {
            headers: None,
            data: Some(serde_json::Value::String("not-base64!!!".to_string())),
        };
        let result: Result<Option<serde_json::Value>> = c.decode(&bad_value);
        assert!(result.is_err());
    }

    // ── encode error produces correct shape ─────────────────────────

    #[test]
    fn encode_error_produces_correct_shape() {
        let err = Error::Application {
            message: "boom".to_string(),
        };
        let encoded = encode_error(&err);
        assert_eq!(encoded["__type"], "error");
        assert_eq!(encoded["message"], "application error: boom");
    }
}

File: impl/sdk/py/src/resonate/codec.py
from __future__ import annotations

import base64
import binascii
import contextlib
import pickle
from typing import TYPE_CHECKING, Any, Protocol

import msgspec

from resonate.error import (
    ApplicationError,
    Base64DecodeError,
    SerializationError,
)
from resonate.types import PromiseRecord, Value

if TYPE_CHECKING:
    from pydantic import BaseModel
else:
    try:
        from pydantic import BaseModel
    except ModuleNotFoundError:  # pydantic is an opt-in extra
        BaseModel = None


def enc_hook(obj: Any) -> Any:
    """Serialize types msgspec can't handle natively (opt-in Pydantic support)."""
    if BaseModel is not None and isinstance(obj, BaseModel):
        return obj.model_dump(mode="json")
    msg = f"cannot encode object of type {type(obj).__name__}"
    raise TypeError(msg)


def dec_hook(expected: type, obj: Any) -> Any:
    """Reshape decoded builtins into a Pydantic model when that's the target."""
    if (
        BaseModel is not None
        and isinstance(expected, type)
        and issubclass(expected, BaseModel)
    ):
        return expected.model_validate(obj)
    msg = f"cannot decode object of type {expected}"
    raise TypeError(msg)


class Encryptor(Protocol):
    def encrypt(self, data: bytes) -> bytes:
        """Encrypts the given byte data."""

    def decrypt(self, data: bytes) -> bytes:
        """Decrypts the given byte data."""


class NoopEncryptor:
    """No-op encryptor (passthrough)."""

    def encrypt(self, data: bytes) -> bytes:
        return data

    def decrypt(self, data: bytes) -> bytes:
        return data


class Codec:
    """Handles encoding/decoding of values for the durability boundary.

    Encode: value -> JSON -> encrypt -> base64 -> ``Value { headers, data }``
    Decode: ``Value { headers, data }`` -> base64 -> decrypt -> JSON -> value
    """

    def __init__(self, encryptor: Encryptor) -> None:
        self.encryptor = encryptor

    def encode(self, value: Any | Exception) -> Value:
        """Encode a serializable value into the wire format.

        An ``Exception`` is always treated as a rejection: it is flattened to
        the error shape by :func:`_encode_error` (a plain ``message`` string,
        plus a best-effort ``__py_pickle`` payload so the awaiter can recover
        the original exception). A resolved value is never an exception object.
        """
        if value is None:
            return Value(headers=None, data=None)

        try:
            json_bytes = msgspec.json.encode(
                _encode_error(value) if isinstance(value, Exception) else value,
                enc_hook=enc_hook,
            )
        except (TypeError, ValueError, msgspec.MsgspecError) as exc:
            raise SerializationError(exc) from exc

        encrypted = self.encryptor.encrypt(json_bytes)
        return Value(headers=None, data=base64.b64encode(encrypted).decode("ascii"))

    def decode(self, value: Value) -> Any:
        """Decode a wire-format value into plain Python builtins.

        Crosses the wire (base64 -> decrypt -> JSON) and stops at builtins;
        returns ``None`` for empty or ``null`` ``data``. Reshaping the decoded
        builtins into a concrete type is :meth:`convert`'s job.
        """
        if not value.data:
            return None

        try:
            data = base64.b64decode(value.data, validate=True)
        except (binascii.Error, ValueError) as exc:
            raise Base64DecodeError(exc) from exc

        try:
            return msgspec.json.decode(self.encryptor.decrypt(data))
        except msgspec.MsgspecError as exc:
            raise SerializationError(exc) from exc

    def convert[T](self, value: Any, type: type[T]) -> T:
        """Coerce an already-decoded value into ``type``.

        Wraps :func:`msgspec.convert` so type-shaping of decoded payloads stays
        inside the codec. Distinct from :meth:`decode`, which crosses the wire
        boundary: this only reshapes a value already on the Python side.
        """
        try:
            return msgspec.convert(value, type, dec_hook=dec_hook)
        except (TypeError, ValueError, msgspec.MsgspecError) as exc:
            raise SerializationError(exc) from exc

    def decode_error(self, value: Value) -> BaseException:
        """Decode a rejected promise's wire ``value`` into its originating error.

        Returns the original exception when its ``__py_pickle`` payload
        round-trips (same Python process / importable class), otherwise an
        :class:`ApplicationError`.
        """
        return _deserialize_error(self.decode(value))

    def decode_promise(self, promise: PromiseRecord) -> PromiseRecord:
        """Decode a promise's ``param`` and ``value`` fields in place.

        Only the two ``Value.data`` payloads change; ``msgspec.structs.replace``
        carries every other field (including each ``Value.headers``) through
        untouched, so the record stays robust to new fields.
        """
        return msgspec.structs.replace(
            promise,
            param=msgspec.structs.replace(
                promise.param, data=self.decode(promise.param)
            ),
            value=msgspec.structs.replace(
                promise.value, data=self.decode(promise.value)
            ),
        )


_PICKLE_KEY = "__py_pickle"


def _encode_error(err: Exception) -> dict[str, str]:
    """Encode an error for durable storage.

    ``message`` is a plain string and is always present. ``__py_pickle`` is a
    best-effort extra field -- a base64 pickle of the original exception so an
    awaiter can recover its exact type and attributes. Pickling can fail
    (unpicklable closures, locks, file handles, ...); on failure the field is
    simply omitted and the awaiter falls back to ``message``.
    """
    encoded = {"__type": "error", "message": str(err)}
    # Best-effort: any pickling failure (unpicklable closures, locks, ...)
    # leaves only ``message``, and the awaiter falls back to ApplicationError.
    with contextlib.suppress(Exception):
        encoded[_PICKLE_KEY] = base64.b64encode(pickle.dumps(err)).decode("ascii")
    return encoded


def _deserialize_error(value: Any) -> BaseException:
    """Deserialize an error value from a rejected promise.

    Prefers the original exception via ``__py_pickle`` when it round-trips
    (same Python runtime, importable class, intact payload); otherwise -- a
    non-Python producer, an unimportable class, a corrupt or absent payload --
    falls back to :class:`ApplicationError` carrying ``message``. The unpickled
    object is type-checked: a payload that does not deserialize to a
    ``BaseException`` is treated as a failure, not raised verbatim.

    Note: ``pickle.loads`` executes arbitrary code from the promise value;
    this trusts the Resonate server and peer workers as the payload source.
    """
    if isinstance(value, dict):
        pickled = value.get(_PICKLE_KEY)
        if isinstance(pickled, str):
            try:
                obj = pickle.loads(base64.b64decode(pickled, validate=True))  # noqa: S301
            except Exception:  # best-effort; fall back to message
                obj = None
            if isinstance(obj, BaseException):
                return obj

        msg = value.get("message")
        if isinstance(msg, str):
            return ApplicationError(msg)

    # msgspec.json.encode outputs bytes, so .decode() is necessary here to format the string
    rendered = msgspec.json.encode(value).decode("utf-8")
    return ApplicationError(f"unknown error: {rendered}")


def decode_settled(record: PromiseRecord) -> Any:
    """Map an already-decoded, settled record to its value, raising on rejection.

    The record's ``value`` has already been decoded by
    :meth:`Codec.decode_promise`, so a resolved payload is returned as-is and
    any rejected payload is turned back into the originating error via
    :func:`_deserialize_error`.
    """
    assert record.state != "pending"
    match record.state:
        case "resolved":
            return record.value.data
        case "rejected" | "rejected_canceled" | "rejected_timedout":
            raise _deserialize_error(record.value.data)
